src/Core/Grid/Query/CombinationQueryBuilder.php

```diff
@@ -0,0 +1,160 @@
+     */
+    public function __construct(
+        Connection $connection,
+        string $dbPrefix,
+        DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator
+    ) {
+        parent::__construct($connection, $dbPrefix);
+        $this->searchCriteriaApplicator = $searchCriteriaApplicator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSearchQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->getCombinationsQueryBuilder($searchCriteria)->select('pa.*');
+
+        $this->searchCriteriaApplicator
+            ->applyPagination($searchCriteria, $qb)
+            ->applySorting($searchCriteria, $qb)
+        ;
+
+        return $qb;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        return $this->getCombinationsQueryBuilder($searchCriteria)
+            ->select('COUNT(pa.id_product_attribute)')
+        ;
+    }
+
+    /**
+     * @param SearchCriteriaInterface $searchCriteria
+     *
+     * @return QueryBuilder
+     */
+    private function getCombinationsQueryBuilder(SearchCriteriaInterface $searchCriteria): QueryBuilder
+    {
+        $filters = $searchCriteria->getFilters();
+        if (!isset($filters['product_id'])) {
+            //@todo: better exception? or require filter in another layer - like builder for filters? (categories has same issue)
+            throw new \RuntimeException('Product id is required for combinations grid');
+        }
+
+        $productId = (int) $filters['product_id'];
+
+        $qb = $this->connection->createQueryBuilder();
+        $qb->from($this->dbPrefix . 'product_attribute', 'pa')
+            ->where('pa.id_product = :productId')
+            ->setParameter('productId', $productId)
+        ;
+
+        // filter by attributes
+        if (isset($filters['attribute_ids'])) {
+            $combinationIds = $this->getCombinationIdsByAttributeIds($productId, (array) $filters['attribute_ids']);
+            $qb->andWhere($qb->expr()->in('pa.id_product_attribute', ':combinationIds'))
+                ->setParameter('combinationIds', $combinationIds, Connection::PARAM_INT_ARRAY)
+            ;
+        }
+
+        if (isset($filters['reference'])) {
+            $qb->andWhere('pa.reference LIKE :reference')
+                ->setParameter('reference', '%' . $filters['reference'] . '%')
+            ;
+        }
+
+        if (isset($filters['default_on'])) {
+            if ((bool) $filters['default_on']) {
+                $qb->andWhere('pa.default_on = 1');
+            } else {
+                $qb->andWhere('pa.default_on IS NULL OR pa.default_on = 0');
+            }
+        }
+
+        return $qb;
+    }
+
+    /**
+     * @param int $productId
+     * @param int[] $attributeIds
+     *
+     * @return int[]
+     */
+    private function getCombinationIdsByAttributeIds(int $productId, array $attributeIds): array
```

Comment:
I'm not sure the productId is required here, since we already filter by the attributeIds anyway Or it could be optional as well if consider my suggestion of not requiring it we would have 

`private function getCombinationIdsByAttributeIds(array $attributeIds, ?int $productId = null): array`

and it would be called like this
`$combinationIds = $this->getCombinationIdsByAttributeIds((array) $filters['attribute_ids'], $filters['product_id'] ?? null);`

Reply:
we want to only get combinations for specified product, but not all the products (its attributeId not the product_attribute_id)